### PR TITLE
Clustering and Service Discovery Support

### PR DESCRIPTION
**Pull request by supriya-premkumar**
Changes:
            1. Adds full cluster support.
            2. Spins up 4 containers. node0, node1, node2 and client
            3. Removes the need for extraneuous ports and uses the following defaults.
                    i.   17000 for ingress coordination svc and 18000 for the corresponding RaftGroup
                    ii.  17001 for ingress shard0 and 18001 for the shard0 RaftGroup
                    iii. 17002 for ingress shard1 and 18002 for the shard1 RaftGroup
            4. All data will live under /pv
            5. Adds logging into /pv/logs
            6. Uses raft-net docker network to enable service discovery.

### CODE REVIEW

File: bootstrap.sh [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  kv -i "$(hostname)" -l :17001 -r "$(hostname)":18001 -d "$(hostname)/shard-0" > /pv/logs/shard0.log 2>&1 &
+  kv -i "$(hostname)" -l :17002 -r "$(hostname)":18002 -d "$(hostname)/shard-1" > /pv/logs/shard1.log 2>&1 &
+
+  echo "Starting Co-ordination Service"
+  kv -i "$(hostname)" -c -l :17000 -r "$(hostname)":18000 -d "$(hostname)" > /pv/logs/coord.log 2>&1
+elif [ -n "$BOOTSTRAP_FOLLOWER" ]; then
+  echo "$(hostname) is configured as follower"
+  echo "Starting Shard Services"
+  sleep 5
+  kv -i "$(hostname)" -l :17001 -r "$(hostname)":18001 -j node0:17001 -d "$(hostname)/shard-0" > /pv/logs/"$(hostname)-shard0.log" 2>&1 &
```
Comment (varks) [inline bootstrap.sh:16]: I do not see `-d` being enabled.  I believe it is still in progress.

**Comment by alex8937** [comment]
qq: I am planning to try our code on [Stanford share computing cluster](https://uit.stanford.edu/service/sharedcomputing). For example, let's say I take 4 nodes and configure as follows

```
Node 1: coordinator 1,  shard1-1, shard3-1, shard4-1, 
Node 2: coordinator 2, shard1-2 shard2-1, shard4-2,
Node 3: coordinator 3,  shard1-3 shard2-2, shard3-2
Node 4: coordinator 4,  shard2-3, shard3-3, shard4-3,
```
So each shard is has 3 replicas and each node serves 3 different shards. The IP address is different for each node. And then I will use goroutine to either on my local machine or other node to launch multiple clients submitting request to test throughput and latency.

 How would this affect that task ?

**Comment by supriya-premkumar** [comment]
> qq: I am planning to try our code on [Stanford share computing cluster](https://uit.stanford.edu/service/sharedcomputing). For example, let's say I take 4 nodes and configure as follows
> 
> ```
> Node 1: coordinator 1,  shard1-1 shard3-1 shard4-1
> Node 2: coordinator 2,  shard1-2 shard2-1 shard4-2
> Node 3: coordinator 3,  shard1-3 shard2-2 shard3-2
> Node 4: coordinator 4,  shard2-3 shard3-3 shard4-3
> ```
> 
> So each shard is has 3 replicas and each node serves 3 different shards. The IP address is different for each node. And then I will use goroutine on either local machine or other node to launch multiple clients submitting request to test throughput and latency.
> 
> How would this affect that task ?

As long as the the nodes are discoverable and you have the updated shard.config you're good.
The only caveat being we use id's as dns names. In other words as long as there is underlying dns resolutions for hostname configured you should be able to port it there.  

Since we don't do auto service discovery. You need to label env vars in your nodes to have `BOOTSTRAP_LEADER` and `BOOTSTRAP_FOLLOWER` set. Depending on how you provision I can help out with that. 

Usually people use an infra provisioner like terraform. Not sure if we get those access on the secure cluster. 

I can help out if you get blocked on this one after you give this a try Yeah?

**Comment by varks** [comment]
> coordinator

nit: let's have 3 coordinators or 5. Let's choose an odd number.

**Comment by varks** [comment]
> qq: I am planning to try our code on [Stanford share computing cluster](https://uit.stanford.edu/service/sharedcomputing). For example, let's say I take 4 nodes and configure as follows
> 
> ```
> Node 1: coordinator 1,  shard1-1 shard3-1 shard4-1
> Node 2: coordinator 2,  shard1-2 shard2-1 shard4-2
> Node 3: coordinator 3,  shard1-3 shard2-2 shard3-2
> Node 4: coordinator 4,  shard2-3 shard3-3 shard4-3
> ```
> 
> So each shard is has 3 replicas and each node serves 3 different shards. The IP address is different for each node. And then I will use goroutine on either local machine or other node to launch multiple clients submitting request to test throughput and latency.
> 
> How would this affect that task ?

nit: let's have 3 coordinators or 5. Let's choose an odd number.

File: common/common.go [hunk 2]
```diff
@@ -43,6 +43,7 @@ const (
 	NodeIDLen           = 5
 
 	MagicDiff = 20000
+	RaftPVBaseDir = "/pv"
```
[the remark this answers is not in the document]
Comment (supriya-premkumar) [reply common/common.go:46]: Here pv refers to PersistedVolume. This is basically for host volume mount.
Working on that.

File: main.go [hunk 3]
```diff
@@ -87,9 +87,8 @@ func main() {
 		if cohortRaftAddress == "" {
 			cohortRaftAddress = common.GetDerivedAddress(raftAddress)
 		}
-		kv := store.NewStore(logger, nodeID, raftAddress, raftDir, joinHTTPAddress == "", listenAddress, bucketName, cohortRaftAddress, joinHTTPAddress)
+		kv := store.NewStore(logger, nodeID, raftDir, raftAddress, joinHTTPAddress == "", listenAddress, bucketName, cohortRaftAddress, joinHTTPAddress)
```
[the remark this answers is not in the document]
Comment (supriya-premkumar) [reply main.go:90]: wanted to group them together. This should not have any impact

File: store/store.go [hunk 4]
```diff
@@ -49,35 +48,37 @@ type Store struct {
 }
 
 // NewStore returns a new Store.
-func NewStore(logger *log.Logger, nodeID, raftAddress, raftDir string, enableSingle bool,
```
[the remark this answers is not in the document]
Comment (supriya-premkumar) [reply store/store.go:52]: should not have any functional difference. We can come back and add this if needed.

**Pull request merged**